src/uct/tcp/sockcm/sockcm_ep.c

```diff
@@ -14,26 +14,215 @@
+        goto out;
+    }
+    ep->sock_id_ctx->handler_added = 0;
+    dest_addr = (struct sockaddr *) &(ep->remote_addr);
+
+    status = ucs_socket_create(dest_addr->sa_family, SOCK_STREAM,
+                               &ep->sock_id_ctx->sock_id);
+    if (status != UCS_OK) {
+        ucs_debug("unable to create client socket for sockcm");
+        goto out_free;
+    }
+
+    ucs_list_add_tail(&iface->used_sock_ids_list, &ep->sock_id_ctx->list);
+    ucs_debug("ep %p, new sock_id %d", ep, ep->sock_id_ctx->sock_id);
+    status = UCS_OK;
+    goto out;
+
+out_free:
+    ucs_free(ep->sock_id_ctx);
+out:
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+    return status;
+}
+
+static inline void uct_sockcm_ep_add_to_pending(uct_sockcm_iface_t *iface, uct_sockcm_ep_t *ep)
+{
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    ucs_list_add_tail(&iface->pending_eps_list, &ep->list_elem);
+    ep->is_on_pending = 1;
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+}
+
+static void uct_sockcm_ep_event_handler(int fd, void *arg)
+{
+    ucs_debug("not implemented");
+}
+
 static UCS_CLASS_INIT_FUNC(uct_sockcm_ep_t, const uct_ep_params_t *params)
 {
-    uct_sockcm_iface_t *iface = ucs_derived_of(params->iface,
-                                               uct_sockcm_iface_t);
+    const ucs_sock_addr_t *sockaddr = params->sockaddr;
+    uct_sockcm_iface_t    *iface    = NULL;
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
 
+    iface = ucs_derived_of(params->iface, uct_sockcm_iface_t);
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
+
+    if (iface->is_server) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (!(params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR)) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    UCT_SOCKCM_CB_FLAGS_CHECK((params->field_mask &
+			       UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS) ?
+			      params->sockaddr_cb_flags : 0);
+
+    self->pack_cb       = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_SOCKADDR_PACK_CB) ?
+                          params->sockaddr_pack_cb : NULL;
+    self->pack_cb_arg   = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_USER_DATA) ?
+                          params->user_data : NULL;
+    self->pack_cb_flags = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS) ?
+                          params->sockaddr_cb_flags : 0;
+    pthread_mutex_init(&self->ops_mutex, NULL);
+    ucs_queue_head_init(&self->ops);
+
+    if (sockaddr->addr->sa_family == AF_INET) {
+        memcpy(&self->remote_addr, sockaddr->addr, sizeof(struct sockaddr_in));
+    } else if (sockaddr->addr->sa_family == AF_INET6) {
+        memcpy(&self->remote_addr, sockaddr->addr, sizeof(struct sockaddr_in6));
+    } else {
+        ucs_error("sockcm ep: unknown remote sa_family=%d", sockaddr->addr->sa_family);
+        status = UCS_ERR_IO_ERROR;
+        goto err;
```

Comment:
maybe it is better to destroy socket ctxs which are failing, rather than leave it to iface destructor?